packages/teleport/src/Apps/useApps.ts

```diff
@@ -15,42 +15,128 @@
+  const { search, pathname } = useLocation();
+  const [startKeys, setStartKeys] = useState<string[]>([]);
   const [isAddAppVisible, setAppAddVisible] = useState(false);
   const { clusterId, isLeafCluster } = useStickyClusterId();
-  const { attempt, setAttempt, run } = useAttempt('processing');
-  const [apps, setApps] = useState([] as App[]);
+  const { attempt, setAttempt } = useAttempt('processing');
   const isEnterprise = ctx.isEnterprise;
+  const [fetchStatus, setFetchStatus] = useState<FetchStatus>('');
+  const [params, setParams] = useState<ResourceUrlQueryParams>({
+    ...getResourceUrlQueryParams(search),
+    sort: { fieldName: 'name', dir: 'ASC' },
+  });
 
-  function refresh() {
-    return ctx.appService
-      .fetchApps(clusterId)
-      .then(res => setApps(res.apps))
-      .catch((err: Error) =>
-        setAttempt({ status: 'failed', statusText: err.message })
-      );
+  const [results, setResults] = useState<AppsResponse>({
+    apps: [],
+    hasResources: false,
+    startKey: '',
+    totalCount: 0,
+  });
+
+  const pageSize = 15;
+
+  const from =
+    results.totalCount > 0 ? (startKeys.length - 2) * pageSize + 1 : 0;
+  const to = results.totalCount > 0 ? from + results.apps.length - 1 : 0;
+
+  useEffect(() => {
+    fetch();
+  }, [clusterId, search]);
+
+  function replaceHistory(path: string) {
+    history.replace(path);
   }
 
   const hideAddApp = () => {
     setAppAddVisible(false);
-    refresh();
+    fetch();
   };
 
   const showAddApp = () => {
     setAppAddVisible(true);
   };
 
-  useEffect(() => {
-    run(() =>
-      ctx.appService.fetchApps(clusterId).then(res => setApps(res.apps))
-    );
-  }, [clusterId]);
+  function setSort(sort: SortType) {
+    setParams({ ...params, sort });
+  }
+
+  function fetch() {
+    setAttempt({ status: 'processing' });
+    ctx.appService
+      .fetchApps(clusterId, { ...params, limit: pageSize })
+      .then(res => {
+        setResults(res);
+        setFetchStatus(res.startKey ? '' : 'disabled');
+        setStartKeys([res.apps[0]?.id, res.startKey]);
+        setAttempt({ status: 'success' });
+      })
+      .catch((err: Error) => {
+        setAttempt({ status: 'failed', statusText: err.message });
+        setResults({ ...results, apps: [], totalCount: 0 });
+        setStartKeys(['']);
+      });
+  }
+
+  const fetchNext = () => {
+    setFetchStatus('loading');
+    ctx.appService
+      .fetchApps(clusterId, {
+        ...params,
+        limit: pageSize,
+        startKey: results.startKey,
+      })
+      .then(res => {
+        setResults({
+          ...results,
+          apps: res.apps,
+          startKey: res.startKey,
+        });
+        setFetchStatus(res.startKey ? '' : 'disabled');
+        setStartKeys([...startKeys, res.startKey]);
+      })
+      .catch((err: Error) => {
+        setAttempt({ status: 'failed', statusText: err.message });
+      });
+  };
+
+  const fetchPrev = () => {
```

Comment:
These fetch funcs are duplicated for each type of resource - I wonder if there's a way we can share more code rather than duplicate.